File: dollar_scraper/download_dollar_bills.py
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.chrome.service import Service
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
from webdriver_manager.chrome import ChromeDriverManager
import os
import time
import requests

def init_driver(headless=False):
    chrome_options = Options()
    if headless:
        chrome_options.add_argument('--headless')
    service = Service(ChromeDriverManager().install())
    driver = webdriver.Chrome(service=service, options=chrome_options)
    return driver

def download_image(url, save_dir, image_number):
    try:
        response = requests.get(url, timeout=30)
        if response.status_code == 200:
            # Use image_number to name the file
            filename = os.path.join(save_dir, f'python_image_{image_number}.jpg')
            with open(filename, 'wb') as f:
                f.write(response.content)
            print(f'Downloaded: {filename}')
    except Exception as e:
        print(f'Failed to download {url}: {e}')

def collect_and_download_images(driver, search_term, num_images, save_dir):
    if not os.path.exists(save_dir):
        os.makedirs(save_dir)
        
    search_url = f'https://www.google.com/search?hl=en&tbm=isch&q={search_term}'
    driver.get(search_url)

    collected_images = 0
    while collected_images < num_images:
        driver.execute_script('window.scrollTo(0, document.body.scrollHeight);')
        time.sleep(2)
        thumbnails = WebDriverWait(driver, 10).until(
            EC.presence_of_all_elements_located((By.CSS_SELECTOR, 'img.Q4LuWd'))
        )
        
        for img in thumbnails[collected_images:num_images]:
            try:
                img.click()
                time.sleep(1)
                images = WebDriverWait(driver, 10).until(
                    EC.presence_of_all_elements_located((By.CSS_SELECTOR, 'img.iPVvYb'))
                )
                for image in images:
                    src = image.get_attribute('src')
                    if src.startswith('http') and src not in save_dir:
                        collected_images += 1
                        # Now pass collected_images as the image number to download_image
                        download_image(src, save_dir, collected_images)
                        if collected_images == num_images:
                            return
            except Exception as e:
                print(f'Error: {e}')

def main():
    driver = init_driver(headless=False)
    dataset_dir = os.path.join(os.getcwd(), 'images/dataset')
    
    try:
        terms = ['5 dollar US bill', '10 dollar US bill', '50 dollar US bill']
        for term in terms:
            print(f'Collecting images for: {term}')
            term_dir = os.path.join(dataset_dir, term.replace(' ', '_'))
            collect_and_download_images(driver, term, 1, term_dir)  # Set the number of images you want to collect
    finally:
        driver.quit()

if __name__ == '__main__':
    main()

File: dollar_scraper/download_dollar_bills.js
const fs = require('fs')
const path = require('path')
const fetch = require('node-fetch')
const puppeteer = require('puppeteer')
// const { setTimeout } = require('node:timers/promises')

const downloadImage = async (url, path, number) => {
    try {
        const response = await fetch(url)
        const buffer = await response.buffer()
        fs.writeFile(path, buffer, () => console.log(`Downloaded ${number}: ${path}`))
    } catch (err) {
        console.log(err.message)
    }
}

const initBrowser = async (headless = true) => {
    const browser = await puppeteer.launch({
        headless,
        args: [
            `--window-size=1480,1024`,
            `--use-fake-ui-for-media-stream`,
            `--enable-features=NetworkService`,
        ],
        defaultViewport: {
            width: 1480,
            height: 1024
        }
    })
    return browser
}

// const collectAndDownloadImages = async (browser, searchTerm, numImages, saveDir) => {
//     if (!fs.existsSync(saveDir)) {
//         fs.mkdirSync(saveDir, { recursive: true })
//     }

//     const page = await browser.newPage()
//     // await page.setViewport({ width: 1920, height: 1080 })
//     await page.goto(`https://www.google.com/search?hl=en&tbm=isch&q=${searchTerm}`)

//     let collectedImages = 0
//     while (collectedImages < numImages) {
//         const thumbnails = await page.$$('img.Q4LuWd')
//         for (const thumb of thumbnails.slice(collectedImages, numImages)) {
//             // await page.evaluate((element) => {
//             //     element.scrollIntoView({ block: 'center', inline: 'center' })
//             // }, thumb)
//             await thumb.click()

//             let src
//             try {
//                 const imgSelector = 'img.iPVvYb'
//                 await page.waitForSelector(imgSelector, { visible: true, timeout: 10000 }).catch(() => {
//                     throw new Error('Full resolution image not found')
//                 })

//                 src = await page.evaluate((sel) => {
//                     const image = document.querySelector(sel)
//                     return image ? image.src : ''
//                 }, imgSelector)

//                 // Check if src is valid and not a placeholder or error indicator
//                 if (!src) {
//                     throw new Error('Invalid image src')
//                 }

//             } catch (error) {
//                 console.log(`${error.message}, trying to download lower resolution`)
//                 src = await thumb.evaluate(el => el.src) // Fallback to the thumbnail src
//             }

//             // Validate src before proceeding
//             if (src && src.startsWith('http')) {
//                 collectedImages += 1
//                 const filename = path.join(saveDir, `image_${collectedImages}.jpg`)
//                 await downloadImage(src, filename, collectedImages)
//                 if (collectedImages === numImages) {
//                     return
//                 }
//             }

//             // await new Promise(resolve => setTimeout(resolve, 1000))
//         }

//         // // Check for and click the 'Show more results' button if not enough images were collected.
//         // if (collectedImages < numImages) {
//         //     const moreResultsButton = await page.$('input[value=\'Show more results\']')
//         //     if (moreResultsButton) {
//         //         await moreResultsButton.click()
//         //         await setTimeout(2000)
//         //     } else {
//         //         await page.evaluate(() => window.scrollTo(0, document.body.scrollHeight))
//         //         await setTimeout(2000)
//         //     }
//         // }
//     }
// }

const collectAndDownloadImages = async (browser, searchTerm, numImages, saveDir) => {
    if (!fs.existsSync(saveDir)) {
        fs.mkdirSync(saveDir, { recursive: true });
    }

    const page = await browser.newPage();
    await page.goto(`https://www.google.com/search?hl=en&tbm=isch&q=${searchTerm}`);

    let collectedImages = 0;
    while (collectedImages < numImages) {
        const thumbnails = await page.$$('img.Q4LuWd');
        for (const thumb of thumbnails.slice(collectedImages, numImages)) {
            // Use page.evaluate to scroll into view and click in the same context
            await page.evaluate(async (element) => {
                // Scroll the thumbnail into view
                element.scrollIntoView({ behavior: 'smooth', block: 'center', inline: 'nearest' });
                // Simulate a click on the thumbnail
                element.click();
                // return fullResImgSrc;
            }, thumb);

            let src
            try {
                const imgSelector = 'img.iPVvYb'
                await page.waitForSelector(imgSelector, { visible: true, timeout: 15000 }).catch(() => {
                    throw new Error('Full resolution image not found')
                })

                src = await page.evaluate((sel) => {
                    const image = document.querySelector(sel)
                    return image ? image.src : ''
                }, imgSelector)

                // Check if src is valid and not a placeholder or error indicator
                if (!src) {
                    throw new Error('Invalid image src')
                }

            } catch (error) {
                console.log(`${error.message}, trying to download lower resolution`)
                src = await thumb.evaluate(el => el.src) // Fallback to the thumbnail src
            }

            // Validate src before proceeding
            if (src && src.startsWith('http')) {
                collectedImages += 1
                const filename = path.join(saveDir, `image_${collectedImages}.jpg`)
                await downloadImage(src, filename, collectedImages)
                if (collectedImages === numImages) {
                    return
                }
            }

            // // Validate src before proceeding
            // if (result && result.startsWith('http')) {
            //     collectedImages += 1;
            //     const filename = path.join(saveDir, `image_${collectedImages}.jpg`);
            //     await downloadImage(result, filename, collectedImages);
            //     if (collectedImages === numImages) {
            //         return;
            //     }
            // } else {
            //     console.log(`Skipping image due to invalid src`);
            // }
        }

        // // Scroll to trigger loading of more thumbnails if needed
        // await page.evaluate(() => window.scrollBy(0, window.innerHeight));
        // await page.waitForTimeout(2000);
    }
};


const main = async () => {
    try {
        const browser = await initBrowser(false)
        const datasetDir = path.join(process.cwd(), 'images/dataset')

        const terms = ['United States five dollar note']
        for (const term of terms) {
            console.log(`Collecting images for: ${term}`)
            const termDir = path.join(datasetDir, term.replace(/ /g, '_'))
            await collectAndDownloadImages(browser, term, 2000, termDir)
        }

        await browser.close()
    } catch (error) {
        console.error('An error occurred:', error)
        process.exit(1) // Exit with an error status code
    }
}

main().catch(error => {
    console.error(error)
    process.exit(1) // Exit with an error status code if there's an unhandled exception
})
